STOR-1726: Add OLM metadata for SMB operator

jsafrane opened the pull request:
All metadata + scripts for the SMB CSI driver operator. This is copy + paste of AWS EFS metadata, I only had to add permissions to create/modify PodDisruptionBudgets that we don't use in EFS.

Tested using:
1. build the operator image + push it somewhere (`quay.io/jsafrane/scratch:smbo1`)
2. (using the upstream driver image, `registry.k8s.io/sig-storage/smbplugin:v1.14.0`):

```
./create-samba-bundle registry.k8s.io/sig-storage/smbplugin:v1.14.0 quay.io/jsafrane/scratch:smbo1 quay.io/jsafrane/scratch:bundle3  quay.io/jsafrane/scratch:index3
```
It will build the bundle + index image, push them to quay and generate CatalogSource on stdout.

3. Apply the generated CatalogSource:
```
oc apply -f - <<EOF
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: smb
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: quay.io/jsafrane/scratch:index3
EOF
```

4. Install the operator in OLM console.

~WIP: audit the permissions~

dobsonj commented on config/samba/bundle.Dockerfile:
```diff
@@ -0,0 +1,9 @@
+LABEL operators.operatorframework.io.bundle.manifests.v1=manifests/
+LABEL operators.operatorframework.io.bundle.metadata.v1=metadata/
+LABEL operators.operatorframework.io.bundle.package.v1=smb-csi-driver-operator
+LABEL operators.operatorframework.io.bundle.channels.v1=stable
+LABEL operators.operatorframework.io.bundle.channel.default.v1=stable
```
Should this channel be `preview` like [secrets-store](https://github.com/openshift/secrets-store-csi-driver-operator/blob/main/config/bundle.Dockerfile#L6) while it's still TechPreview?
If yes, then it also means renaming config/manifests/stable to config/manifests/preview and `s/stable/preview/` in the other files too.

jsafrane replied:
good point, moved to `preview/`

dobsonj commented on config/samba/manifests/stable/smb-csi-driver-operator.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,369 @@
+    repository: https://github.com/openshift/csi-operator
+    createdAt: "2021-07-14T00:00:00Z"
+    description: Install and configure CIFS/SMB CSI driver.
+    olm.skipRange: ">=4.9.0-0 <4.16.0"
```
```suggestion
    olm.skipRange: ">=4.15.0-0 <4.16.0"
```

dobsonj commented on config/samba/manifests/art.yaml:
```diff
@@ -0,0 +1,18 @@
+    # replace entire version line, otherwise would replace 4.3.0 anywhere
+    - search: "version: {MAJOR}.{MINOR}.0"
+      replace: "version: {FULL_VER}"
+    - search: 'olm.skipRange: ">=4.9.0-0 <{MAJOR}.{MINOR}.0"'
+      replace: 'olm.skipRange: ">=4.9.0-0 <{FULL_VER}"'
```
```suggestion
    - search: 'olm.skipRange: ">=4.16.0-0 <{MAJOR}.{MINOR}.0"'
      replace: 'olm.skipRange: ">=4.16.0-0 <{FULL_VER}"'
```

dobsonj commented on config/samba/manifests/stable/image-references:
```diff
@@ -0,0 +1,29 @@
+apiVersion: image.openshift.io/v1
+spec:
+  tags:
+  - name: smb-csi-driver-rhel8-operator
```
```suggestion
  - name: smb-csi-driver-rhel9-operator
```

dobsonj commented on config/samba/manifests/stable/smb-csi-driver-operator.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,369 @@
+    - name: Source Repository
+      url: https://github.com/openshift/csi-operator
+  version: 4.16.0
+  maturity: stable
```
```suggestion
  maturity: preview
```

dobsonj commented on config/samba/manifests/stable/smb-csi-driver-operator.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,369 @@
+  maintainers:
+    - email: aos-storage-staff@redhat.com
+      name: Red Hat
+  minKubeVersion: 1.21.0
```
```suggestion
  minKubeVersion: 1.29.0
```

dobsonj commented on hack/opm-bundle/bundle.Dockerfile:
```diff
@@ -0,0 +1,9 @@
+FROM scratch
```
Do you really want these files duplicated under `hack/opm-bundle`? We don't do this for the other OLM operators. I added a [.gitignore line](https://github.com/openshift/secrets-store-csi-driver-operator/blob/main/.gitignore#L19) for secrets-store (but it's the only one with an explicit ignore).

jsafrane replied:
good catch, yes, they should be ignored

jsafrane commented:
I checked permissions in the operator CSV file and moved bunch of them from `clusterrules` to `rules` and removed some unnecessary ones (e.g. CloudCredentialsRequest). I consider them completeWIP: >.

jsafrane commented:
/retest-required

dobsonj commented:
/lgtm
/label docs-approved
/label px-approved
/label qe-approved

dobsonj commented:
/override ci/prow/e2e-azure-ovn-upgrade
This job passed before tide retested everything, and it really has nothing to do with the OLM manifests in this PR.
